File: ScriptGraph/TemplateScriptNode/EyebrowRaise/CGEyebrowRaise.js
/**
 * @file CGEyebrowRaise.js
 * @author liujiacheng
 * @date 2021/8/18
 * @brief CGEyebrowRaise.js
 * @copyright Copyright (c) 2021, ByteDance Inc, All Rights Reserved
 */

const {BaseNode} = require('./BaseNode');
const Amaz = effect.Amaz;

class CGEyebrowRaise extends BaseNode {
  constructor() {
    super();
    this.actionDetected = false;
  }

  onUpdate(sys, dt) {
    const algMgr = Amaz.AmazingManager.getSingleton('Algorithm');
    const algResult = algMgr.getAEAlgorithmResult();
    let faceIdx = this.inputs[0]();

    if (algResult !== null && faceIdx !== null) {
      const face = algResult.getFaceBaseInfo(Math.round(faceIdx));
      if (face === undefined || face === null) {
        return;
      }

      const hasAction = face.hasAction(Amaz.FaceAction.BROW_JUMP);
      if (!this.actionDetected && hasAction) {
        if (this.nexts[0]) {
          this.nexts[0]();
        }
      } else if (this.actionDetected && hasAction) {
        if (this.nexts[1]) {
          this.nexts[1]();
        }
      } else if (this.actionDetected && !hasAction) {
        if (this.nexts[2]) {
          this.nexts[2]();
        }
      } else {
        if (this.nexts[3]) {
          this.nexts[3]();
        }
      }
      this.actionDetected = hasAction;
    }
  }
}

exports.CGEyebrowRaise = CGEyebrowRaise;

File: ScriptGraph/TemplateScriptNode/EyebrowRaise/CGEyebrowRaise.lua
local CGEyebrowRaise = CGEyebrowRaise or {}
CGEyebrowRaise.__index = CGEyebrowRaise

function CGEyebrowRaise.new()
    local self = setmetatable({}, CGEyebrowRaise)
    self.inputs = {}
    self.nexts = {}
    return self
end

function CGEyebrowRaise:setNext(index, func)
    self.nexts[index] = func
end

function CGEyebrowRaise:setInput(index, func)
    self.inputs[index] = func
end

function CGEyebrowRaise:update()
    if self.nexts[0] == nil then
        return
    end

    local result = Amaz.Algorithm.getAEAlgorithmResult()
    local face = result:getFaceBaseInfo(self.inputs[0]())
    if face == nil then
        return
    end

    local is_BROW_JUMP = face:hasAction(Amaz.FaceAction.BROW_JUMP)
    if is_BROW_JUMP then
        if self.nexts[0] then
            self.nexts[0]()
        end
    end
end

return CGEyebrowRaise

